couchdb/server/views.py

```diff
@@ -0,0 +1,56 @@
+# -*- coding: utf-8 -*-
+#
+import copy
+import logging
+
+from couchdb import json
+from couchdb.server.exceptions import QueryServerException, Error
+
+__all__ = ('map_doc',)
+
+log = logging.getLogger(__name__)
+
+
+def map_doc(server, doc):
+    """Applies available map functions to document.
+
+    :command: map_doc
+
+    :param server: Query server instance.
+    :type server: :class:`~couchdb.server.BaseQueryServer`
+
+    :param doc: Document object.
+    :type doc: dict
+
+    :return: List of key-value results for each applied map function.
+
+    :raises:
+        - :exc:`~couchdb.server.exceptions.Error`
+          If any Python exception occurs due mapping.
+    """
+    docid = doc.get('_id')
+    log.debug('Apply map functions to document `%s`:\n%s', docid, doc)
+    orig_doc = copy.deepcopy(doc)
+    map_results = []
+    _append = map_results.append
+    try:
+        for idx, func in enumerate(server.state['functions']):
+            # TODO: https://issues.apache.org/jira/browse/COUCHDB-729
+            # Apply copy.deepcopy for `key` and `value` to fix this issue
+            _append([[key, value] for key, value in func(doc) or []])
```

Comment:
Hmm... An edge case here. Do we need to take care it?

> [[key, value] for key, value in func(doc) or []]

If user make `func` as a generator, but `yield` nothing, the following evaluation will happen:
1. `func(doc) or []`
2. `<generator object> or []`
3. `<generator object>` win!
4. doing args unpacking on `None`
5. Boom!

```diff
diff --git a/couchdb/tests/server/views.py b/couchdb/tests/server/views.py
index eab81a0..fe31d05 100644
--- a/couchdb/tests/server/views.py
+++ b/couchdb/tests/server/views.py
@@ -117,6 +117,15 @@ class MapTestCase(unittest.TestCase):
         doc = {'_id': 'foo'}
         views.map_doc(self.server, doc)
 
+    def test_yield_nothing(self):
+        state.add_fun(
+            self.server,
+            'def mapfun(doc):\n'
+            '  yield'
+        )
+        doc = {'_id': 'foo'}
+        views.map_doc(self.server, doc)
+
 
 class ReduceTestCase(unittest.TestCase):
```

output:
```python
======================================================================
ERROR: test_yield_nothing (couchdb.tests.server.views.MapTestCase)
shouldn't crush if map function yield nothing
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/home/iblis/git/couchdb-python/couchdb/tests/server/views.py", line 128, in test_yield_nothing
    views.map_doc(self.server, doc)
  File "/usr/home/iblis/git/couchdb-python/couchdb/server/views.py", line 54, in map_doc
    raise Error(err.__class__.__name__, str(err))
Error: ('TypeError', "'NoneType' object is not iterable")

----------------------------------------------------------------------
```